File: src/services/audit/detectors/ad/groups/size/excessive-members.ts
/**
 * GROUP_EXCESSIVE_MEMBERS - Groups with excessive direct members
 */

import { ADGroup } from '../../../../../../types/ad.types';
import { Finding } from '../../../../../../types/finding.types';
import { toAffectedADGroupEntities } from '../../../../../../utils/entity-converter';

/**
 * Detect groups with excessive direct members
 * Large groups are difficult to manage and review
 */
export function detectGroupExcessiveMembers(groups: ADGroup[], includeDetails: boolean): Finding {
  const EXCESSIVE_THRESHOLD = 100;

  const affected = groups.filter((g) => {
    const memberCount = g.member?.length ?? 0;
    return memberCount > EXCESSIVE_THRESHOLD;
  });

  // Sort by member count (most members first)
  affected.sort((a, b) => (b.member?.length ?? 0) - (a.member?.length ?? 0));

  return {
    type: 'GROUP_EXCESSIVE_MEMBERS',
    severity: 'medium',
    category: 'groups',
    title: 'Group with Excessive Members',
    description: `Groups with more than ${EXCESSIVE_THRESHOLD} direct members. Large groups are difficult to audit and may grant unintended access.`,
    count: affected.length,
    affectedEntities: includeDetails ? toAffectedADGroupEntities(affected) : undefined,
    details:
      affected.length > 0
        ? {
            threshold: EXCESSIVE_THRESHOLD,
            largestGroups: affected.slice(0, 5).map((g) => ({
              name: g.sAMAccountName || g.dn,
              memberCount: g.member?.length ?? 0,
            })),
            recommendation:
              'Review large groups and consider breaking into smaller, role-based groups for better access control.',
          }
        : undefined,
  };
}

File: internal/audit/detectors/ad/groups/size/excessive-members.go
package size

import (
	"context"
	"sort"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// ExcessiveMembersDetector checks for groups with excessive direct members
type ExcessiveMembersDetector struct {
	audit.BaseDetector
}

// NewExcessiveMembersDetector creates a new detector
func NewExcessiveMembersDetector() *ExcessiveMembersDetector {
	return &ExcessiveMembersDetector{
		BaseDetector: audit.NewBaseDetector("GROUP_EXCESSIVE_MEMBERS", audit.CategoryGroups),
	}
}

const excessiveThreshold = 100

// Detect executes the detection
func (d *ExcessiveMembersDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	type groupInfo struct {
		name        string
		memberCount int
	}

	var affected []groupInfo

	for _, group := range data.Groups {
		memberCount := len(group.Member)
		if memberCount > excessiveThreshold {
			name := group.SAMAccountName
			if name == "" {
				name = group.DistinguishedName
			}
			affected = append(affected, groupInfo{name: name, memberCount: memberCount})
		}
	}

	// Sort by member count (most members first)
	sort.Slice(affected, func(i, j int) bool {
		return affected[i].memberCount > affected[j].memberCount
	})

	var affectedNames []string
	for _, g := range affected {
		affectedNames = append(affectedNames, g.name)
	}

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityMedium,
		Category:    string(d.Category()),
		Title:       "Group with Excessive Members",
		Description: "Groups with more than 100 direct members. Large groups are difficult to audit and may grant unintended access.",
		Count:       len(affected),
	}

	if len(affected) > 0 {
		finding.AffectedEntities = toAffectedGroupEntities(affectedNames)

		// Build largest groups list (max 5)
		largestGroups := make([]map[string]interface{}, 0, 5)
		for i := 0; i < len(affected) && i < 5; i++ {
			largestGroups = append(largestGroups, map[string]interface{}{
				"name":        affected[i].name,
				"memberCount": affected[i].memberCount,
			})
		}

		finding.Details = map[string]interface{}{
			"threshold":      excessiveThreshold,
			"largestGroups":  largestGroups,
			"recommendation": "Review large groups and consider breaking into smaller, role-based groups for better access control.",
		}
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewExcessiveMembersDetector())
}
